### PR TITLE
Logfile public examples

### PR DESCRIPTION
**Pull request by abstractmachines**

### CODE REVIEW

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+#include "LogfileTestNode.hpp"             // For all CPP Logfile API examples.
+
+using namespace std;
+
+using namespace polysync::datamodel;
+
+
+void LogFileTestNode::prepareLogfileToIterate()
+{
+    sleepMicro ( 1000000 ); //always sleep before modeoff so queue doesn't flush.
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: modeoff --> mode off

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    cout<<"\n\n************************************************\n"
+    "*** PolySync LogFile API: C++ Iterator Example ***\n\n";
+
+    try
+    {
+        LogFileTestNode aNode;
+
+        sleep( 2 );
+
+// Regarding setting node name using Logfile Iterator and how that relates to
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: I'd prefer these to be block comments.

/\* 
 * Regarding setting node name using Logfile Iterator and how that relates to
 * setFilePath(), setSessionId(), and setNodeName():
 * 
 * Example: in directory 1234, let a logfile be name.5678.plog,
 * where 1234 = session ID, 5678 = GUID, and name = node name.
 * 
 * To iterate over records in a logfile, use its explicit filepath.
 * 
 * Any file specified in the path can be read from; node name is irrelevant.
 * How: use any setNodeName(), and use setFilePath() in prepareLogfileToIterate().
 */

File: LogfileWriter/PolySyncLogfile.cpp [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+    {
+        cout <<"\nLog Record: Index: " << logRecord->index <<" - Size: "
+             << logRecord->size <<" - Previous Size: "
+             << logRecord->prev_size <<" - RnR Timestamp ( \nheader.timestamp):"
+             << logRecord->timestamp;
+    }
+
+    cout << endl;
+
+    // get ps_msg_ref, and do something with it.
```
Comment (shnewto) [inline LogfileWriter/PolySyncLogfile.cpp]: is this referring to something I should see?

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+{
+/**** Logfile Iterator. No code should be in in okStateEvent(). ****/
+}
+
+
+void LogFileTestNode::releaseStateEvent()
+{
+    printResults();
+
+// Turn off mode. Turning off the mode automatically disables state.
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: indent these comments?

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    cout<<"\nReleasing logfile resources. If all messages did not print "
+          "to Terminal \nin either read, write, or iterator, that is due to "
+          "i/o (printf / cout) \nbeing slower than CPU.\n\n";
+}
+
+
+void LogFileTestNode::initStateEvent()
+{
+// 1. Init LogFile API resources:
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: indent these comments?

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    {
+        cout<<"Log Record: Index: " << logRecord->index <<" - Size: "
+           << logRecord->size <<" - Previous Size: "
+           << logRecord->prev_size <<" - RnR Timestamp ( \nheader.timestamp):"
+           << logRecord->timestamp;
+    }
+
+    cout<<endl;
+
+    // get ps_msg_ref, and do something with it.
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: Not sure what this refers to?

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    if( logRecord )
+    {
+        cout<<"Log Record: Index: " << logRecord->index <<" - Size: "
+           << logRecord->size <<" - Previous Size: "
+           << logRecord->prev_size <<" - RnR Timestamp ( \nheader.timestamp):"
+           << logRecord->timestamp;
+    }
+
+    cout<<endl;
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: missing a space

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 1 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/e4d88fc026ab64982608c5850d9955a01cbef5bc

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 2 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/e4d88fc026ab64982608c5850d9955a01cbef5bc

File: LogfileWriter/PolySyncLogfile.cpp (hunk 3 above)
Comment (abstractmachines) [inline LogfileWriter/PolySyncLogfile.cpp]: Taken from C Logfile example. Removed! Thanks :) https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/e4d88fc026ab64982608c5850d9955a01cbef5bc

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 4 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/e4d88fc026ab64982608c5850d9955a01cbef5bc

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 5 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/e4d88fc026ab64982608c5850d9955a01cbef5bc

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 6 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: Taken from C Logfile example. Removed! https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/e4d88fc026ab64982608c5850d9955a01cbef5bc

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 7 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/e4d88fc026ab64982608c5850d9955a01cbef5bc

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+{
+    sleepMicro ( 1000000 ); // Sleep before mode off so queue doesn't flush.
+
+    _logFile->setModeOff(); // Automatically disables state.
+
+    _logFile->setFilePath( "/tmp/polysync_logfile.plog"  );
+
+    cout << "Logfile Iterator started." << endl;
+
+    _logFileWasIterated = true;
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: Seems like this should be false to start right? or maybe the name is misleading me?

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        const ps_rnr_log_record * const logRecord,
+        void * const userData )
+{
+    cout << "\nLogfile iterator callback\n";
+
+    if( logRecord )
+    {
+        cout<<"Log Record: Index: " << logRecord->index <<" - Size: "
+           << logRecord->size <<" - Previous Size: "
+           << logRecord->prev_size <<" - RnR Timestamp ( \nheader.timestamp):"
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: newline in string here?

File: LogfileIterator/LogfileIteratorExample.cpp [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+
+void LogFileTestNode::releaseStateEvent()
+{
+    printResults();
+
+    /* Turn off mode. Turning off the mode automatically disables state.
+     * Sleep before turning mode off after last write to avoid flushing of queue.
+     */
+    sleep( 5 );
```
Comment (shnewto) [inline LogfileIterator/LogfileIteratorExample.cpp]: what's the reason you're not using sleepMicro calls here? The psync call plays nicely and you're using it in other places.

File: LogfileIterator/PolySyncLogfile.cpp [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+
+void Logfile::setStateEnabled( ps_timestamp offsetToStart )
+{
+    setState( LOGFILE_STATE_ENABLED, offsetToStart );
+}
+
+
+void Logfile::setStateDisabled()
+{
+    setState( LOGFILE_STATE_DISABLED, 1000 );
```
Comment (shnewto) [inline LogfileIterator/PolySyncLogfile.cpp]: What's the magic number represent?

File: LogfileIterator/PolySyncLogfile.cpp [hunk 12]
```diff
@@ -0,0 +1,501 @@
...
+    }
+    return false;
+}
+
+
+void Logfile::readerResetToBeginning()
+{
+    checkDTC( psync_logfile_reader_reset_to_beginning( _node ),
+                "Failed to reset LogFile Reader to beginning of file." );
+}
```
Comment (shnewto) [inline LogfileIterator/PolySyncLogfile.cpp]: does the reset function need to set any flags to false/true? I'm imagining the "was iterated" flag is a candidate.

File: LogfileQueueReader/LogfileQueueReaderExample.cpp [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+    //    filterOutMessages();
+
+    // 2. Choose Read method: either filepath (simplest use case), or session ID.
+    // Option 1: set filepath. If set by writer in same file, no need to reset.
+    _logFile->setFilePath( "/tmp/polysync_logfile.plog"  );
+
+    _logFile->setSessionId( 1 );
+
+    // Option 2: Using setSessionId() instead of setFilePath():
+    //_logFile->setSessionId( 1234 );
```
Comment (shnewto) [inline LogfileQueueReader/LogfileQueueReaderExample.cpp]: How does this an Option 2 that makes it different from the setSessionId( 1 ) used above?

File: LogfileQueueReader/PolySyncLogfile.cpp [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+
+    checkDTC( psync_logfile_reader_get_eof_status
+              ( _node, &eofReached ),
+                "Failed to get LogFile EOF status." );
+
+    if ( eofReached )
+    {
+        return true;
+    }
+    return false;
```
Comment (shnewto) [inline LogfileQueueReader/PolySyncLogfile.cpp]: need a newline before the return statement.

File: LogfileWriter/LogfileWriterExample.cpp [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+     * ps_timestamp stateTimingOffsetDistributedContext =
+     *      getTimestamp() + 2000000;
+     *
+     * _logFile->setStateEnabled( stateTimingOffsetDistributedContext );
+     */
+
+    // 5. Set flags / counters.
+    _numMessagesWritten = 0;
+
+    _messagesWereWritten = true;
```
Comment (shnewto) [inline LogfileWriter/LogfileWriterExample.cpp]: I'm not sure why the prepare functions set the "were written" flags to true?

File: LogfileWriter/LogfileWriterExample.cpp [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+{
+    /* For each okStateEvent() loop, process a single write + sleep (required).
+     * For larger messages such as Lidar, set a larger (higher) write frequency.
+     * Terminate with Ctrl + C.
+     */
+    // 1. Write a message.
+    writeMessage();
+
+    // 2. Set write frequency for 1x Replay.
+    polysync::sleepMicro( 5000 );
```
Comment (shnewto) [inline LogfileWriter/LogfileWriterExample.cpp:167]: you're already using the polysync namespace right? no need for polysync::?

File: LogfileWriter/PolySyncLogfile.cpp (hunk 3 above)
Comment (shnewto) [inline LogfileWriter/PolySyncLogfile.cpp]: Looks like it hasn't been removed yet?

Comment (abstractmachines) [inline LogfileWriter/PolySyncLogfile.cpp]: https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/4459cbb9d677792ef6e926bdda92d006acffea2b

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 8 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: Artifacts remaining from previous Logfile Example work setup. Fixed / removed this logic altogether in all Logfile Examples. https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/4459cbb9d677792ef6e926bdda92d006acffea2b

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 9 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: The C Logfile Iterator example does not have a newline here. The CPP Logfile API emulates the C Logfile API as closely as possible.

File: LogfileIterator/LogfileIteratorExample.cpp (hunk 10 above)
Comment (abstractmachines) [inline LogfileIterator/LogfileIteratorExample.cpp]: Fixed! Good point. https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/4459cbb9d677792ef6e926bdda92d006acffea2

File: LogfileIterator/PolySyncLogfile.cpp (hunk 11 above)
Comment (abstractmachines) [inline LogfileIterator/PolySyncLogfile.cpp]: The Logfile documentation for the API can be pretty lengthy. Part of that documentation is as such: when enabling state for single node context, timing delay is zero. When enabling state for distributed context, timing delay is now + 2 seconds, generally.

When disabling state and/or setting state to invalid, a near-zero time delay is preferable. This is  a near-zero time delay here that's the "magic number." This is documented in the Logfile API. Near-zero is preferable over zero for timing skew and timing delay.

File: LogfileIterator/PolySyncLogfile.cpp (hunk 12 above)
Comment (abstractmachines) [inline LogfileIterator/PolySyncLogfile.cpp]: This function wraps the C Logfile API which returns DTC's. For the CPP Logfile Example purposes, this function is never invoked.

File: LogfileQueueReader/LogfileQueueReaderExample.cpp (hunk 13 above)
Comment (abstractmachines) [inline LogfileQueueReader/LogfileQueueReaderExample.cpp]: That does seem unclear doesn't it? Improved the documentation https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/4459cbb9d677792ef6e926bdda92d006acffea2b

File: LogfileQueueReader/PolySyncLogfile.cpp (hunk 14 above)
Comment (abstractmachines) [inline LogfileQueueReader/PolySyncLogfile.cpp]: Done. https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/4459cbb9d677792ef6e926bdda92d006acffea2b

File: LogfileWriter/LogfileWriterExample.cpp (hunk 15 above)
Comment (abstractmachines) [inline LogfileWriter/LogfileWriterExample.cpp]: Removed. Artifact from previous Example architecture.

File: LogfileWriter/LogfileWriterExample.cpp (hunk 16 above)
Comment (abstractmachines) [inline LogfileWriter/LogfileWriterExample.cpp:167]: Removed. Thanks https://github.com/PolySync/PolySync-CPP-Examples/pull/13/commits/4459cbb9d677792ef6e926bdda92d006acffea2b

**Pull request closed without merging**